### PR TITLE
adds ability to create keys from crypto.PrivateKey types

### PR DESCRIPTION
**Pull request by adam-hanna**

### CODE REVIEW

File: crypto/key.go [hunk 1]
```diff
@@ -174,6 +180,42 @@ func GenerateEKeyPair(curveName string) ([]byte, GenSharedKey, error) {
 	return pubKey, done, nil
 }
 
+// KeyPairFromKey generates a new private and public key from an input private key
+func KeyPairFromKey(priv crypto.PrivateKey) (PrivKey, PubKey, error) {
+	if priv == nil {
+		return nil, nil, ErrNilPrivateKey
+	}
+
+	switch priv.(type) {
```
Comment (Stebalien) [inline crypto/key.go]: use `switch priv := priv.(type)` to save the result. That way, you don't need to cast in all the cases.

File: crypto/key.go [hunk 2]
```diff
@@ -174,6 +180,42 @@ func GenerateEKeyPair(curveName string) ([]byte, GenSharedKey, error) {
...
+	case *rsa.PrivateKey:
+		log.Println("1")
```
Comment (Stebalien) [inline crypto/key.go]: debugging code.

File: crypto/key.go [hunk 3]
```diff
@@ -174,6 +180,42 @@ func GenerateEKeyPair(curveName string) ([]byte, GenSharedKey, error) {
 	return pubKey, done, nil
 }
 
+// KeyPairFromKey generates a new private and public key from an input private key
+func KeyPairFromKey(priv crypto.PrivateKey) (PrivKey, PubKey, error) {
```
Comment (Stebalien) [inline crypto/key.go]: This will need to work when `-tags=openssl` is passed.

File: crypto/key.go (hunk 2 above)
Comment (adam-hanna) [reply crypto/key.go]: ack, embarrassing!

File: crypto/key.go (hunk 3 above)
Comment (adam-hanna) [reply crypto/key.go]: Done

File: crypto/key.go (hunk 1 above)
Comment (adam-hanna) [reply crypto/key.go]: done

**Comment by adam-hanna** [comment]
Thanks for the help!

**Comment by adam-hanna** [comment]
Scratch this, this isn't working...

`// +build !openssl` makes my code unable to find `KeyPairFromKey`

**Comment by Stebalien** [comment]
@raulk this extends the core interface slightly but seems useful (and looks correct).

File: crypto/key_not_openssl.go [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+	"crypto"
+	"crypto/ecdsa"
+	"crypto/rsa"
+
+	btcec "github.com/btcsuite/btcd/btcec"
+	"golang.org/x/crypto/ed25519"
+)
+
+// KeyPairFromKey generates a new private and public key from an input private key
+func KeyPairFromKey(priv crypto.PrivateKey) (PrivKey, PubKey, error) {
```
Comment (raulk) [inline crypto/key_not_openssl.go]: How about we name this `KeyPairFromStdKey` to more accurately describe what’s going on?

File: crypto/key_not_openssl.go [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+import (
+	"crypto"
+	"crypto/ecdsa"
+	"crypto/rsa"
+
+	btcec "github.com/btcsuite/btcd/btcec"
+	"golang.org/x/crypto/ed25519"
+)
+
+// KeyPairFromKey generates a new private and public key from an input private key
```
Comment (raulk) [inline crypto/key_not_openssl.go]: Godoc is a bit bleak. “Generates” is an imprecise term here, as there’s no key generation happening, rather wrapping or conversion.

File: crypto/key_not_openssl.go (hunk 4 above)
Comment (adam-hanna) [reply crypto/key_not_openssl.go]: done

File: crypto/key_not_openssl.go (hunk 5 above)
Comment (adam-hanna) [reply crypto/key_not_openssl.go]: done

**Comment by adam-hanna** [comment]
Thanks @raulk 

Done!

**Comment by Stebalien** [comment]
Thanks!

**Pull request merged**